### DISCUSSION TITLE
How to reduce the time taken by MagickWand methods

### DISCUSSION THREAD

**Discussion by programmerarunkumar** [ImageMagick/ImageMagick#1885] [General]
I'm using MagickWand methods to resize image
MagickResizeImage() ,MagickReadImage()  taking more time.
Is there any way to reduce the time taken ?

Time Taken : https://drive.google.com/file/d/1dHumBlDo2S4giS2eQc6cgFCFbI-7FNWA/view

```C
// ----My Code-----
#include <stdio.h>
#include <stdlib.h>
#include <MagickWand/MagickWand.h>

int main(int argc,char **argv)
{   
  MagickWand *magick_wand;
  int width,height;
  clock_t start,end;
  clock_t read_start,read_end;
  clock_t resize_start,resize_end;
  clock_t write_start,write_end;
  FILE *fp;
  if (argc != 3)
    {
      (void) fprintf(stdout,"Usage: %s image thumbnail\n",argv[0]);
      exit(0);
    }

  start = clock();
  fp = fopen(argv[1], "r");
  if (fp == NULL)
      printf("\nFile unable to open ");

  fseek(fp, 0, 2);  
  int size = ftell(fp);
  printf("The size of the file : %d MB\n", size/(1024*1024));

  
  MagickWandGenesis();

  magick_wand=NewMagickWand(); 
 

  read_start=clock();
  MagickReadImage(magick_wand,argv[1]);
  read_end=clock();

  resize_start=clock();	
  MagickResizeImage(magick_wand,1000,1000,LanczosFilter);
  resize_end=clock();
  
  write_start=clock();
  MagickWriteImages(magick_wand,argv[2],MagickTrue);  
  write_end=clock();	
  
  magick_wand=DestroyMagickWand(magick_wand);

  MagickWandTerminus();

  end=clock();
  double total_time = (double)(end - start) / CLOCKS_PER_SEC;
  double read_time = (double)(read_end - read_start) / CLOCKS_PER_SEC;
  double resize_time = (double)(resize_end - resize_start) / CLOCKS_PER_SEC;
  double write_time = (double)(write_end - write_start) / CLOCKS_PER_SEC;
  printf("The program took %f seconds to read , %f seconds to resize , %f seconds to write , %f seconds for total execution\n", read_time,resize_time,write_time,total_time);
  return(0);
}
```

**Comment by dlemstra** [comment]
Which version of ImageMagick are you using?

**Reply by programmerarunkumar** [reply]
7.0.9

**Reply by dlemstra** [reply]
That is not the complete version. Are you using a `HDRI` build? Disabling `HDRI` could improve the performance. And using `Q8` instead of `Q16` could also help.

**Reply by programmerarunkumar** [reply]
> That is not the complete version. Are you using a `HDRI` build? Disabling `HDRI` could improve the performance. And using `Q8` instead of `Q16` could also help.

How to disable HDRI  & How to use Q8

**Comment by urban-warrior** [comment]
Want more performance from ImageMagick? Try these options:

- add more memory to your system, see the [pixel cache](https://imagemagick.org/script/architecture.php#cache);
- add more cores to your system, see [threads of execution support](https://imagemagick.org/script/architecture.php#threads);
- reduce lock contention with the [tcmalloc](http://goog-perftools.sourceforge.net/doc/tcmalloc.html) memory allocation library;
- push large images to a solid-state drive, see [large image support](https://imagemagick.org/script/architecture.php#tera-pixel).

If these options are prohibitive, you can reduce the quality of the image results. The default build is Q16 HDRI. If you disable [HDRI](https://imagemagick.org/script/high-dynamic-range.php), you use half the memory and instead of predominately floating point operations, you use the typically more efficient integer operations. The trade-off is reduced precision and you cannot process out of range pixel values (e.g. negative). If you build the Q8 non-HDRI version of ImageMagick, you again reduce the memory requirements in half-- and once again there is a trade-off, even less precision and no out of range pixel values. For a Q8 non-HDRI build of ImageMagick, use these configure script options: `--with-quantum-depth=8 --disable-hdri`.

**Reply by programmerarunkumar** [reply]
Getting The requested topic does not exist. for the above given link
https://drive.google.com/file/d/1TinrG6bVR9nMC-4eIwgkGnZQC0kYwuMg/view?usp=sharing

**Reply by urban-warrior** [reply]
We replaced the link with appropriate text.

**Comment by snibgo** [comment]
The Q-number and HDRI choice are options in "./configure". At the bash command line, type "./configure --help" for a list of options.

Alternatives to resize include scale and sample. These are faster and satisfactory for some purposes.

**Comment by programmerarunkumar** [comment]
I don't want to compromise in the quality of the resized image and aslo want to reduce the time taken for resizing the image considerable
What is the best option to achieve it ?

**Comment by snibgo** [comment]
> What is the best option to achieve it ?

urban-warrior has suggested four options that won't reduce the quality. The "best" option will depend on where the bottleneck lies. For example, if the command involves almost no disk activity but CPUs are saturated, then get a faster CPU.

**Comment by programmerarunkumar** [comment]
> Alternatives to resize include scale and sample. These are faster and satisfactory for some purposes.

Can We expect the same quality from scale and sample as same as resize ?

**Reply by fmw42** [reply]
Each is different.  resize interpolates between pixels to get good quality.  Scale does a simple average of blocks of pixels. Sample skips every so many pixels to keep the original colors.  The former two will change or introduce new colors. So you have to choose which you want depending upon your needs. Quality is a subjective matter. It is a trade between speed with possible artifacts and introduction of new colors and precision. What one person likes may not suit someone else.
